### DISCUSSION TITLE
Transitioning Arduino Code to AutoPilot

### DISCUSSION THREAD

**Discussion by jmdelahanty** [wehr-lab/autopilot#150] [Q&A] [answered]
Hello fellow autopilots!

Our lab is considering moving into the wonderful world of `autopilot` for running a distributed set of behavior boxes used for training mice. We currently run via Arduinos which works, but they're all controlled by a single MATLAB instance and the boxes cannot be run independently of one another. Autopilot seems like a perfect solution to this problem!

We've ordered a couple Pi's to test with that will arrive in the coming months, so while we wait I figured it'd be a good time to start porting what was on our Arduino's to Autopilot!

[Here's a link](https://github.com/Tyelab/bruker_control/blob/main/bruker_disc_specialk/bruker_disc_specialk.ino) to the Arduino code that we'd like to port over. ~The training boxes don't require much of the functionality that's in it (i.e. stimulations, catch trials) so it'll hopefully be simpler.~

EDIT: It would actually be cool to have those things enabled in the training in case someone wants to do that in the boxes. Might as well do it anyways...

The way the Arduino runs is that a `Python` program called `bruker_control` generates a pseudorandom trial list that follows certain rules. From there, specific arrays are generated for ITI timings, tone durations, and stimulation offset timings. These arrays are transmitted to the Arduino over a serial connection and then the Arduino iterates through the trials until the specified number is complete. The configuration that was used with precise timings is then written in a `JSON` format to disk so the exact parameters can be rerun again by someone else if they'd like with literally the same trial set and also so the experimental procedure is documented precisely for each run and programmatically accessible by anyone (autopilot inspired!).

I've been confused by trying to use [PyDAQmx](https://pythonhosted.org/PyDAQmx/) on my own so we could avoid using Arduinos and run experiments exclusively through Python, but got stuck having a camera record video independently of the routine for sending stimuli to the mouse. I was thinking I could basically use the same structure as the arrays that are sent through to the Arduino when sending things via PyDAQmx and have a while loop that updates new timings and checks against flags like the Arduino does for when to send different stimuli, but it didn't seem like that's how people do it in Python as much or, if it is how people do it, I haven't seen that many examples yet.

How hard do you think it might be, fellow pilots, to port this Arduino functionality to `autopilot`? Can I use the same code that generates trial structures, use similar configuration files, and store things in directory structures as we currently do?

**Comment by sneakers-the-rat** [comment, marked as answer]
Hello Jeremy! Can you post some examples of the arrays that you're describing here? maybe give an example play-by-play of "X value used by Y hardware, then Z value used by.." 

It should be very possible to write a Task that, given some parameterization like a list of possible ITIs/etc. to generate them like in the `bruker_control` function (or take whatever it outputs), and the trial timings can be saved in whatever output data you generate from the task, however you'd like it. Shouldn't be too hard!

**Reply by jmdelahanty** [reply]
Here's an example of a real configuration file (I'm not sure if the lab wants to continue outputting these files... EDIT: I think it makes sense to have them so experiments are self documenting so you know precisely what happened without having to load/analyze/clean up the raw data that happens) that was generated when an experiment was completed:
```
{"beh_metadata": {"totalNumberOfTrials": 60, 
"startingReward": 3, 
"maxSequentialReward": null, 
"maxSequentialPunish": 3, 
"punishTone": 2000, 
"rewardTone": 9000,
"USDeliveryTime_Sucrose": 200,
"USDeliveryTime_Air": 100,
"USConsumptionTime_Sucrose": 3000,
"vacuum": true,
"ITIJitter": true,
"baseITI": null,
"minITI": 15,
"maxITI": 30,
"toneJitter": false,
"baseTone": 3,
"minTone": null,
"maxTone": null,
"catchTrials": true,
"numCatchReward": 3,
"numCatchPunish": 2,
"catchOffset": 0.25,
"percentPunish": 0.25,
"stim": true,
"shutter_only": false,
"stimFrequency": 20,
"stimPulseTime": 5,
"stimLambda": 590,
"stimDeliveryTime_PreCS": 1000,
"stimDeliveryTime_Total": 5000,
"stimStartPosition": 20,
"numStimReward": 5,
"numStimPunish": 5,
"numStimAlone": 3,
"trialArray": [1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 4, 6, 5, 5, 4, 4, 5, 6, 5, 4, 6, 4, 5, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 0, 0, 1, 3, 3, 1, 2, 1, 3, 1],
"ITIArray": [27786, 21672, 18075, 21733, 22276, 28625, 29069, 25498, 21512, 23109, 19457, 25481, 17746, 29085, 23783, 19458, 16897,
23362, 23021, 24183, 21297, 17558, 15666, 25821, 21360, 28661, 18414, 22131, 17316, 25218, 29083, 19338, 26272, 27871, 20624, 17682,
26825, 19093, 18319, 17247, 15989, 29162, 20667, 29511, 25161, 26373, 21882, 26154, 15895, 29512, 20942, 25475, 25788, 21308, 22136,
15626, 26422, 20577, 24428, 16171],
"toneArray": [3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000,
3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000,
3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000, 3000],
"LEDArray": [20297, 16558, 14666, 24821, 20360, 27661, 17414, 21131, 16316, 24218, 28083, 18338, 25272],
"dropped_frames": []}, 
"zstack_metadata": {"zstack": true, "stack_number": 3, "zdelta": 25.0, "zstep": 1.0}, "weight_check": true}
```

The relevant arrays are the ones at the bottom: `trialArray`, `ITIArray`, `toneArray`, and `LEDArray`. Most of the values define rules for trial sets or describe parameters used during the experiment. The reason I make one for each session is so the relevant pieces of the parameters used for the behavioral experiment can all be accessed easily programmatically and so there was a standard template for each project to generate trials flexibly in one system without relying on the Arduino (which we're not as familiar with coding in) to do it for you.   It also makes it easier to update where the system can have new functionality made available to anyone who wants to use it by just adding a configuration parameter to things. It also allows for transmissions of metadata that are used for running things and lets you change a file instead of having to go in and update/make individual Arduino scripts for a given change that can't be programmatically accessed easily later. Basically, things are hardcoded less and the Arduino only knows how to receive trials, iterate through them, and then reset itself for the next batch.

The experiment is governed by the `trialArray` and the trial types within. This number defines what type of trial is being executed and, therefore, what types of stimuli are delivered. Each trial has an ITI, so the stimulus is not delivered until the period defined by the ITI (in ms) has elapsed. Nearly every trial has a `tone` in MHz that's delivered. The only trials that have no `tone` are trial-type 6, which is when only an LED pulse is delivered to the mouse. If the trial-type is [4, 5, 6] then a value out of the LED array is selected, which defines the time that must elapse during an ITI before the LED stimulation starts.

Each of these are iterated through as needed until the end of the experiment occurs, when the number of executed trials reaches the total number of trials!

If there's a better way to do this then we're all ears!

**Reply by sneakers-the-rat** [reply]
Sure - the way you would do this in autopilot would be to declare things like `punishTone` et al as `PARAMS` for a task (see an example in the docs: https://docs.auto-pi-lot.com/en/latest/guide/task.html ). 

The rest seems like a distinction between whether you want to have the exact series of events prespecified as some hardcoded array beforehand vs. generating them programmatically by the task class. In the former case, you could also specify them as `PARAM`s, and in the latter you would specify them as `TrialData` and returning them as data from the task, which would also satisfy desires for easy access as it would just be something like `Subject('subject_id').get_trial_data()['trialType']` or whatever you want to call it.

So if you want to do it the hardcoded way, you could pass a a list of trial types, say we save them as `self.trialTypes`, and then have some incrementing counter like

```python
self.trial_num = next(self.counter)
self.trialType = self.trialTypes[self.trial_num]
if self.trialType == 1:
    pass
elif self.trialType == 2:
   pass
# and so on
```

Generating those programmatically would just mean doing something like
```python
self.trialType = random.randint(1,6)
if self.trialType == 1:
  pass
# etc. same as above
```

I think that the latter is easier since you don't need to generate a new hardcoded event series every time, and since the goal of autopilot is to produce data that is 'clean' at the time of acquisition by giving you control over what is stored and how, and since the whole thing is Python you don't really have the same restrictions that make the former necessary in the arduino model.

**Reply by sneakers-the-rat** [reply]
Here's an example of a timed transition between trial stages:
https://github.com/auto-pi-lot/autopilot-plugin-wehrlab/blob/2ba8c23b131c00b477e73ae74b760eb11f0b126f/tuning_curve/tuning_curve.py#L131-L133

in that case there's just one stage that repeats over and over again, but you could eg. have another stage method that starts by playing the sound. If you want really precise ITIs (<~2ms) there are a few other strategies we could try.

**Reply by jmdelahanty** [reply]
Interesting! Thanks Jonny!

> I think that the latter is easier since you don't need to generate a new hardcoded event series every time,

That makes sense. The main reason for having these hardcoded events was so I could have the trial generation that follows specific rules done in Python so the Arduino didn't have to do anything for creating trial events (because I'm bad at Arduino lol).

I wasn't sure how to generate trials one by one that made sure the rules my labmates needed were followed (like no more than 3 trials of each type in a row, make trials that have stimulations or not, catch trials or not, etc). So instead I make a whole new array of just one trial type and then flip trials in specific positions according to where the user wants certain types of trials to happen. I finally check the arrays to make sure the random flips followed the rules.

It basically programmatically generates the `trialArray` at the very beginning for use later. Doing this one at a time with classes was something I thought about and thought would be great to do, but I didn't know how to do it! [Here's](https://github.com/Tyelab/bruker_control/blob/main/main/trial_utils.py) the module that does it. I fear that it's pretty bad, but I did my best to document things.

If I wanted to do the second method, which sounds like a much better way to do it overall, I'd imagine that I could use those rules to specify when those stages should be invoked. Here's some rules that people required that I would try making into `autopilot` stages:

1. The first several trials must all be reward trials: This could be the first stage and it ends according to when a `PARAM` states it, like the `startingReward` field in the configuration JSON.
2. "Normal" stage with pseudorandom generation of trials where rewards and punishments are delivered, a method of the stage would be to flip a trial if a counter for max numbers of sequential trials reaches the `PARAM` `maxSequentialReward`/`maxSequentialPunish`
3. If stimulations trials are required, that's a stage that's used depending on if `PARAM` `stim` is `True`. Then that stage creates trials according to it's own `PARAMS`
4. "Normal stage" again
5. Catch trials stage that's used if `PARAM` `catchTrials` is true.

If I didn't do it this way and ended up using the first method for whatever reason, would each trial effectively be it's own stage?

When it comes to the training piece, having stages for the mice depending on if they've graduated to the next level is an amazing idea that one of our lab's graduate students has been doing already, but it would be a seamless process in `autopilot` which would be awesome.

**Reply by sneakers-the-rat** [reply]
Definitely doable.

You could do this all with a single set of stages (ie. stages just for the different phases of the task like stim presentation, reward delivery, etc.) with conditional logic within them.

So for example with 1) you could make a `n_initial_reward` param, and then use an `itertools.counter` to count trials (as in https://github.com/wehr-lab/autopilot/blob/e408c08e76df8d8c930edfb4ba58e31cd1ec4d87/autopilot/tasks/nafc.py#L298 ). Then you could have a `do_reward` attribute that you at the beginning of the trial you set to `True` if `n_trial` < `n_initial_reward` . 

You could store the recent history of trials in attrs of the task class as well, say you make a `reward_history = collections.deque(maxlen=history_length)`, and then `append` to it, then do something similar to above like `if sum(reward_history[-3:-1]) == 3` then set reward to False, or whatever you need to do.

If you wanted to do the first way, then you could pass a list of trial configuration options, something like
```python
trial_params = [
    {
        'reward': True,
        'stim': :True,
        'other_param': '...'
    },
    {
        # second trial params
    }
]
```

and then iterate through them. That would mean that you would need to regenerate them and the protocol every time if you wanted them to change, otherwise they would be the same every time you ran the mice. 

Re: Graduation -- there are some rules already for accuracy, n_trials, etc. but you can add your own! All you'd need to do is make a class that inherits from `autopilot.tasks.graduation.Graduation`, and make sure that it has some `update` method that takes a `tables.Row` object and returns a `bool` that indicates whether the subject should graduate to the next level in the protocol. The registry should be able to find it if it's in your plugins folder (which is where all your task code should live, see https://docs.auto-pi-lot.com/en/latest/guide/plugins.html )

**Reply by jmdelahanty** [reply]
It'll take us a little while to start implementing this properly, so you'll have to hang tight while we hack away!

In the meantime, one thing is that people only want certain numbers of each trial type overall. I'm assuming that doing this all with the `PARAM`s all set correctly is also very possible.

One thing I was curious about: is there a disadvantage in generating the trials and checking them before iterating through them? I'm wondering if it's best practice to programmatically generate trials during runtime instead of before runtime. And then do you typically store the resulting run as it's own `.json`? I think it'd be nice to keep a record that describes the exact parameters of the experiment (meaning trial orders, stimuli durations, etc) like I do now if I could.

**Reply by sneakers-the-rat** [reply]
yes, very possible to have a parameter that controls number of trials. See my above comment on how to control numbers of types of trials, and if you wanted to stop after a total number of trials you would just count them and when you wanted to stop just do `self.running.clear()` from the task class.
see:
https://github.com/wehr-lab/autopilot/blob/e408c08e76df8d8c930edfb4ba58e31cd1ec4d87/autopilot/core/pilot.py#L781-L783

I think the only disadvantage to generating trial lists beforehand (ie. in a `protocol.json` file) is that it's sorta pointless and more code/labor intensive and a lil brittle, but use it however you want to! One way to do that would be to have each day's trials in their own step, and then have an `n_trials` graduation criteria between them so the mouse graduated at the end of every day. Otherwise ya just parameterizing it in the abstract like having a `n_trials_of_type_whatever` parameter and generating them within the task class once it starts makes more sense to me. 

In either case you will have a record that describes the exact parameters of the experiment, including trial orders, stimulus durations, etc. in the output data. The only difference is whether they are precomputed and stored in the `protocol.json` *and* the output data.

I think we might be talking past each other a bit so here's an example:

I have a protocol like this, where I describe one level of the protocol like this:

```json
  {
    "graduation": {
      "type": "accuracy",
      "value": {
        "threshold": "0.99",
        "window": "400",
        "type": "accuracy"
      }
    },
    "reward": "20",
    "req_reward": false,
    "correction": true,
    "bias_mode": false,
    "punish_dur": "5000",
    "task_type": "2AFC",
    "step_name": "speech_3",
    "pct_correction": "20",
    "stim": {
      "sounds": {
        "R": [
          {
            "token": "0",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/bo/bo_0.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "1",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/bo/bo_1.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "2",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/bo/bo_2.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "0",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/bi/bi_0.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "1",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/bi/bi_1.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "2",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/bi/bi_2.wav",
            "consonant": "b",
            "type": "Speech",
            "amplitude": "0.02"
          }
        ],
        "L": [
          {
            "token": "0",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/go/go_0.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "1",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/go/go_1.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "2",
            "speaker": "Brynna",
            "vowel": "o",
            "path": "Brynna/CV/go/go_2.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "0",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/gi/gi_0.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "1",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/gi/gi_1.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          },
          {
            "token": "2",
            "speaker": "Brynna",
            "vowel": "i",
            "path": "Brynna/CV/gi/gi_2.wav",
            "consonant": "g",
            "type": "Speech",
            "amplitude": "0.02"
          }
        ]
      }
    },
    "punish_sound": false
  }

```

So I have some description of the stimuli I want to present, their parameterization, some stuff that controls trial order (eg. `"correction": true` turns on a bias correction mode where if the mouse gets an incorrect answer on one side, then with some `pct_correction` probability it won't present a stimulus on that side until the mouse guesses the other side), stimulus duration, etc. So i'm not explicitly describing the order of the trials in the protocol, ya?

I can then load the data from a mouse running this protocol (actually a different but similar one, but just for example)

```python
>>> from autopilot.core.subject import Subject
>>> sub = Subject('9523')
>>> sub.get_trial_data().tail()
step groups:
['S07_speech_5']
                        DC_timestamp                   RQ_timestamp  bailed  correct  correction response  session target  ...  consonant                                         path    speaker  token       type vowel step     step_name
12093  b'2019-12-03T15:16:54.737844'  b'2019-12-03T15:16:54.437572'       0        1           1     b'R'       18   b'R'  ...       b'b'  b'/usr/autopilot/sounds/Brynna/ba/ba_1.wav'  b'Brynna'    1.0  b'Speech'  b'a'    7  S07_speech_5
12094  b'2019-12-03T15:16:55.941498'  b'2019-12-03T15:16:55.634164'       0        1           0     b'R'       18   b'R'  ...       b'b'  b'/usr/autopilot/sounds/Brynna/bi/bi_1.wav'  b'Brynna'    1.0  b'Speech'  b'i'    7  S07_speech_5
12095  b'2019-12-03T15:16:57.240616'  b'2019-12-03T15:16:56.909504'       0        0           0     b'R'       18   b'L'  ...       b'g'  b'/usr/autopilot/sounds/Brynna/gi/gi_1.wav'  b'Brynna'    1.0  b'Speech'  b'i'    7  S07_speech_5
12096  b'2019-12-03T15:17:05.228799'  b'2019-12-03T15:17:04.813340'       0        0           0     b'L'       18   b'R'  ...       b'b'  b'/usr/autopilot/sounds/Brynna/bi/bi_0.wav'  b'Brynna'    0.0  b'Speech'  b'i'    7  S07_speech_5
12097  b'2019-12-03T15:17:14.140202'  b'2019-12-03T15:17:13.571462'       0        0           0     b'L'       18   b'R'  ...       b'b'  b'/usr/autopilot/sounds/Brynna/bi/bi_1.wav'  b'Brynna'    1.0  b'Speech'  b'i'    7  S07_speech_5
```

So I have all the details from each trial,  which stimulus was presented, which side the mouse guessed on, in what order, and so on. make sense?

**Reply by jmdelahanty** [reply]
This makes lots of sense! Thanks for the explanation, this is great.

It makes sense that it's more code/labor intensive for sure (I see that myself when adding things to the trial generation!). Could you say a little more about what makes it brittle? I was hoping that the way we've tried having things was robust so I don't want a brittle system!

Another thing that people in the lab want to do is have "Yoked" trials, where we can deliver the same exact trial set for an entire group of animals. Basically once a trial set is generated, people want to use the same thing for the next animals. Is this something we can do as well? It's another reason that having the hardcoded sets appeared helpful since I didn't know how to ensure the trial sets were generated programmatically the exact same way.

**Reply by sneakers-the-rat** [reply]
sort of depends on the exact way it's written, but some examples: say you have some notion of a `trial_type` like `(1, 2, 3, 4)`. You then need some switch case or other way of unpacking that into the relevant parameters. So then adding a new trial type needs you to add a new entry to your switch case and any other potential changes elsewhere in the code. That can get complicated pretty quick. Or, say you have all the separate parameters given in big long lists. Then it can be sort of hard to encapsulate different sets of parameters (eg. how all the sound parameters are grouped together in the above example). What if you want one of the parameters to eventually be constant? do you then write exceptions for all the trial-varying parameters if they are unchanged, or just have to pass one value over and over again? One way of handling that would be to have defaults, but that means you're eventually building down to a point where you can provide one or many values for each of the parameters. Unless you're careful about splitting this all up across a class hierarchy,  all the specificity of how to handle all the lists of parameters could make the code hard to reuse. 

The reason you describe would be the reason you'd want to give specific lists of precomputed parameters: ie. not just the abstract structure of the task, but the order itself is important to the experiment, which is reasonable. In that case yes! you would want to have it in the protocol definition. That's sort of the logical distinction between the `protocol.json` file and the `Task` class. One of them describes the structure of what should be true for all the animals within an experiment, and the other describes how it should be done. Usually the thing we want to have control over is some parameter like "mean number of trials per minute" or "change this parameter slowly over the course of an hour" -- more abstract than the literal values, so the protocol.json is to give control of those.  If exact order is important, to generate the task you then end up writing code outside of the task class in a protocol generating function, which is fine, but then you then have logic split across two types of objects and so it could get tricky to use ideas from one in the other, etc. 

All those depend on exactly how you write it, so it could be very elegant or very messy! just trying to answer the question of "how could that be brittle." I'm not sure exactly why you'd want to do exactly the same order of trials for a group of animals, normally i would assume the randomness is important to isolate the particularity of the stimuli/trials from their order, but i'm sure there's a good reason! so if that's what you need then by all means do that.

**Reply by sneakers-the-rat** [reply]
another thing you could do is set the seed of the random number generator, and depending on how that's done (eg. what sources of entropy it uses, like a time-based one wouldn't produce the same numbers if the tasks were run at different times) that could be another way of doing it. But i think generating lists of parameters sounds easier than that.

**Reply by jmdelahanty** [reply]
> adding a new trial type needs you to add a new entry to your switch case and any other potential changes elsewhere in the code.

This is precisely what happens! I guess what I've been banking on is that, at the moment, it's not too hard to add just a couple lines of code for each new trial type, but it does make things more complicated/harder to implement and maintain.

> Unless you're careful about splitting this all up across a class hierarchy, all the specificity of how to handle all the lists of parameters could make the code hard to reuse.

Class hierarchy has been scary to me, but I gotta get into it sometime!

>  I'm not sure exactly why you'd want to do exactly the same order of trials for a group of animals, normally i would assume the randomness is important to isolate the particularity of the stimuli/trials from their order, but i'm sure there's a good reason! so if that's what you need then by all means do that.

My understanding is that it makes it easier to compare data across animals if they all have the same trial set for a given session. Although it's the same across animals, the subjects themselves will still experience the stimuli as random/unpredictable. Well, in our case pseudo-randomly since we have these rules. Since the data we're collecting typically has a lower number of animals overall in a paper and the number of cells may also be small, we can increase the size of comparable trials for an entire group by doing this "yoked" trial set. The mice can also only be headfixed for a little while and we need to get as much data as possible out of it, so there's randomized ITIs and the like, but we basically have about half an hour to get the data we need out. That puts a fair number of constraints upon the kind of experiments we can do, at least that's my understanding of it.

EDIT: When I was first making trial arrays, ITIs, etc I had things done purely randomly, but the lab decided to have certain trial rules to ensure we get enough data from subjects for each trial type.

**Reply by sneakers-the-rat** [reply]
I'd think especially in a low n setting you would want randomization, otherwise you wouldn't be able to separate "effect of x stimulus in itself" from "effect of x stimulus in this particular recent history of stimuli," but that's a statistical rather than a programming question. if you want it done that way then it can be done that way.

**Reply by jmdelahanty** [reply]
I'm probably misremembering/misrepresenting the idea for why, that's just what I remember. We could maybe talk about that sometime, it would be great to learn your thoughts about that.

Last potentially dumb question (for now!): the Pi4s' pinouts seem to have many types of GPIO pins from regular GPIO pins, MISO/MOSI pins, etc

Are any and all GPIO pins usable for simply sending stimuli and recording voltage outputs and the different types of pins simply have additional capabilities? I just want to make sure that the "general purpose" part of the acronym is still true. And do people record just rising voltage events for timestamps or sample continuously during the session?

**Reply by sneakers-the-rat** [reply]
anything labeled with a 'GPIO' can be used like a GPIO, but only the, eg. 'SDA/SCL' pins can serve as SDA or SCL pins for i2c (not a big deal since i2c is daisy-chainable and register-based). with pigpio we can do hardware PWM on all GPIOs.

not sure what you mean using GPIO to send stimuli, they can do pulses and stuff but i wouldn't like play sound out of them. 

you can sample however you want -- if you want to sample events continuously, see the `record` arg and the `events` attr for `Digital_In`, and then you can save that by just streaming events back to the terminal like any other continuous data.

**Reply by jmdelahanty** [reply]
> not sure what you mean using GPIO to send stimuli, they can do pulses and stuff but i wouldn't like play sound out of them.

I didn't realize how unclear that was lol, I meant send pulses out. Using speakers is something that we need to do though! All we're playing is a higher pitch tone at 9kHz and a lower pitch at 2kHz for the mice. I know that you have examples in your code and docs for how to use it so I will check those out too!

> you can sample however you want -- if you want to sample events continuously, see the record arg and the events attr for Digital_In, and then you can save that by just streaming events back to the terminal like any other continuous data

I'm not sure which one is best to do! I think really all we need is timestamps of events starting/stopping which you've already shown here how to do, but I feel like people having the "raw" data sampled out to a file is something they'd want in this lab. Maybe I can do both, that would be neat.

**Reply by jmdelahanty** [reply]
Something I just realized I needed to ask that's relevant here: Ideally what's running behavior stimulation through the scope is through something like Autopilot. But the machine that runs the scope's software is Windows and there's no Linux distro.

Do you think implementing these things in an Autopilot like fashion is something we can do? Would we invoke these things as a subprocess that goes on while the scope is taking pictures and opencv writes the video to disk?

**Reply by sneakers-the-rat** [reply]
I'm not sure what you mean here, you mean like you have one computer that's running like scanimage or something on windows? Typically I would expect you would be running the behavioral experiment part on a raspi (raspiOS or similar)? To coordinate with a separate windows machine you could communicate with network messages (zmq), serial, or just with a ttl pulse, but I don't think I really understand the question. Maybe a block diagram of the hardware setup you have in mind?

**Reply by jmdelahanty** [reply]
Here's what we currently have going:
![bruker_stim_diagram_current](https://user-images.githubusercontent.com/28411193/157553318-5ca183d6-cfb4-4be2-9382-274008d17b31.png)

And here's what I had meant to say:
![bruker_stim_diagram_future](https://user-images.githubusercontent.com/28411193/157553412-f9bdb8a6-5786-4647-89a1-e0fd075afb76.png)

The way things effectively work right now is basically:

1. Python does the setting up: Gets the scope/scope's software ready, generates trials/ITIs/tone durations/etc and sends everything to the Arduino.
2. The Arduino kicks everything off by sending a `Go` signal to the microscope. All the data that is written to disk is collected according to the Windows' machine clock that's governing the microscope's software.

The thinking behind that was that it would be best if all the data was recorded by one machine and united to the same software so timestamps would all be the same across each datastream. I messaged with Talmo briefly and he suggested doing an interpolation on every datastream's timestamps to put them into a common time base. He also suggested we gather TTL pulses for every time the camera takes an image to have external timestamps. I used to do that, but I only have 7 DAQ channels available and we're pushing that as it is now. Also, since the camera takes a picture only when the microscope takes a picture, it was decided that gathering that information wasn't necessary (for better or worse). This is the part that would be changing the most.

3. Python collects all the video frames specified by the length of the recording (there's a number of frames needed calculated in `video_utils`). Once that's reached, Python sends a kill signal to the scope's API and ends the experiment.

**Reply by sneakers-the-rat** [reply]
Oh yeah totally -- that's how we're using it in a few experiments atm.

tasks are started by the Terminal by sending a `'START'` message with a task description to a pilot object, that looks something like this

```python
from autopilot.networking.node import Net_Node

node = Net_Node(...)
node.send(to='pilot_id', key='START', value=task_params)
```

and then the pilot runs it. See https://github.com/wehr-lab/autopilot/blob/e408c08e76df8d8c930edfb4ba58e31cd1ec4d87/autopilot/core/pilot.py#L271 and https://github.com/wehr-lab/autopilot/blob/e408c08e76df8d8c930edfb4ba58e31cd1ec4d87/autopilot/core/pilot.py#L723

Typically that's done from a Terminal GUI instance, but there's nothing really special about it except that currently the pilot has an unfortunately hardcoded return address for the data as `'T'`, but that too would be pretty easy to replace (and that is the plan). So you could just send a message using a `Net_Node` class named `'T'` and it would pretty much be the same (though you'd have to mimic its handling of the `Subject` class to store data. working on that at the moment actually).

re: synchronization, TTL pulses are always useful as redundancy. Another thing you could do is set the main windows machine to be an NTP server and then setup the pi to synchronize its system time off it, which should be ~us precise over a quiet local network. 

One thing you could do if you need to double up TTL pulses on one channel is just to invert the polarity of one of the inputs so it goes negative (or attenuate one so you have two pulse heights, idk)
